### ISSUE TITLE
Missing property error on a resource-dependent output

### ISSUE THREAD

**Issue by StefWTW** [pulumi/pulumi#5071]
I am trying to create an Azure Function App and assigning its system identity RBAC to a resource.

I get the following error:

```
error: azure:authorization/assignment:Assignment resource 'roleAssignment' has a problem: Missing required property 'principalId'
```

If I run the deployment without the role assignment and then, subsequently include the resource again, it works.

Looks like it doesn't work out the fact that it needs to create the function and its identity first and then use the result for the assignment.

This is my code:

```csharp
var function =
    new FunctionApp("functionApp", new FunctionAppArgs
    {
        [...]
        Identity = new FunctionAppIdentityArgs
        {
            Type = "SystemAssigned"
        }
    });

// This only works the second time
// before the function is deployed, it complains
// about principal ID not being present
new Assignment("roleAssignment", new AssignmentArgs
{
    PrincipalId = function.Identity.Apply(fi => fi.PrincipalId!),
    [...]
});
```

I have also tried to make the dependency explicit with `DependsOn`, but it did not help.

thank you.

**UPDATE**
The second time I tried, after destroying and re-creating the resources it didn't work even removing the resource and adding it later.

**Comment by leezen** [comment]
This reads like this upstream issue: https://github.com/terraform-providers/terraform-provider-azurerm/issues/4532

I'd suggest trying the workaround described in https://github.com/terraform-providers/terraform-provider-azurerm/issues/4532#issuecomment-626877283 to basically do a `get` to lookup the ID and pass that in instead

**Comment by StefWTW** [comment]
@leezen the ID is what I am after, I won't be able to use it to look it up.

in that example they use the display name (which is not ideal, but I guess works as a workaround for Terraform), but in Pulumi there is not `Get` overload that supports looking up by display name.

**Comment by leezen** [comment]
The equivalent to the read that's used in the TF example is https://www.pulumi.com/docs/reference/pkg/azuread/getserviceprincipal/#using

Alternatively, since Pulumi is 'just code' you can also use the Azure SDK to do the lookup as well.

**Comment by StefWTW** [comment]
Thanks, I'll give that a try with the AzureAD provider

**Comment by StefWTW** [comment]
I still get an error `Grpc.Core.RpcException: Status(StatusCode=Unknown, Detail="invocation of azuread:index/getServicePrincipal:getServicePrincipal returned an error: invoking azuread:index/getServicePrincipal:getServicePrincipal: A Service Principal with the Display Name "functionappa3123db1" was not found")`, I presume the SP is not yet visible and that may need a delay to work.

**Comment by StefWTW** [comment]
It looks like it knows the name upfront, therefore tries to get the SP before the function gets created. Is there a way to force the execution of the Get to only after the creation of the resource?

**Comment by NickTolhurst26** [comment]
For what it's worth I just stumbled across this post when dealing with a similar issue. 

It guided me to a 'workaround', here's my snippet:

```csharp
var app = new Azure.AppService.AppService("app", new Azure.AppService.AppServiceArgs
{
    Location = resourceGroup.Location,
    ResourceGroupName = resourceGroup.Name,
    AppServicePlanId = plan.Id,
   
    // ... other config ...
});

var appGet = Azure.AppService.AppService.Get("appget", app.Id, new Azure.AppService.AppServiceState
{
    ResourceGroupName = resourceGroup.Name,
    Location = resourceGroup.Location,
    AppServicePlanId = plan.Id,
},
new CustomResourceOptions
{
    DependsOn = app
});

var roleAssignment = new Azure.Authorization.Assignment("roleAssignment", new Azure.Authorization.AssignmentArgs
{
    Scope = Output.Create(Azure.Core.GetSubscription.InvokeAsync()).Apply(subscription => subscription.Id),
    RoleDefinitionName = "Owner",
    PrincipalId = appGet.Identity.Apply(sp => sp.PrincipalId),
    SkipServicePrincipalAadCheck = true
});
```

I don't like it, but I wanted to avoid using AzureAD, and it did the trick! 

Hope this is useful! :D

**Comment by StefWTW** [comment]
thank you @NickTolhurst26, I need to give that a try. Even if it's hacky it's less hacky than my approach.

I manage to get it to work by using:

```csharp
var principalId =
    Output.Tuple(function.Name, function.Id)
            .Apply(tuple => GetServicePrincipal.InvokeAsync(new GetServicePrincipalArgs
                            {
                                DisplayName = tuple.Item1
                            }, new InvokeOptions { Parent = function }))
            .Apply(sp => sp.ObjectId);
```

So, basically, to force it to wait, I did `Output.Tuple(...., function.Id)` as the ID is something that can't be figured out upfront without Azure. That forced the execution to be delayed until after the resource creation. It's awful and probably won't work in a lot of cases, so would be nice to have a proper solution.

**Comment by leezen** [comment]
Until https://github.com/pulumi/pulumi/issues/4600 happens, running via the `apply` is probably the easiest way to do this. Going to close the issue since the lifecycle stuff is tracked in the other issue and the property issue here I believe has been solved.
